### PR TITLE
Enable function sections with windows-gnu

### PR DESCRIPTION
**Pull request by mati865**
Passed tests locally on x86_64 with LLVM assertions enabled.

### CODE REVIEW

File: src/librustc_target/spec/windows_gnu_base.rs [hunk 1]
```diff
@@ -66,8 +66,7 @@ pub fn opts() -> TargetOptions {
     );
 
     TargetOptions {
-        // FIXME(#13846) this should be enabled for windows
-        function_sections: false,
+        function_sections: true,
```
Comment (petrochenkov) [inline src/librustc_target/spec/windows_gnu_base.rs:69]: The line can be just removed because `true` is the default.

Comment (mati865) [reply src/librustc_target/spec/windows_gnu_base.rs]: Right, thanks.

**Comment by nagisa** [comment]
IIRC the reason it was disabled until now was because on Windows there's an egregiously small limit on sections in object files, which then leads to compilation failures. Is `windows-gnu` not affected by this?

**Comment by mati865** [comment]
Apparently it was disabled because of the LLVM assertions: https://github.com/rust-lang/rust/issues/13846
LLD supports doing it for both MSVC and GNU Windows targets.

I have tried enabling it some time ago but got ui tests crashes.
Just recently I have noticed enabling ASLR or functions_sections had very similar (or even identical) crashes as one of my mistakes when working on https://github.com/rust-lang/rust/pull/72049. To be exact the crashes happened when statics were imported from other DLLs without dllimport.
If you want I can verify tomorrow if this is unexpected "fallout" from https://github.com/rust-lang/rust/pull/72049 or something else has fixed it.

**Comment by mstorsjo** [comment]
> IIRC the reason it was disabled until now was because on Windows there's an egregiously small limit on sections in object files, which then leads to compilation failures. Is `windows-gnu` not affected by this?

FWIW, the limit on sections in an object file is higher if compiling in "big object" mode. IIRC GCC and/or MSVC require you to pass specific flags for enabling this, while LLVM enables this format automatically as needed.

**Comment by mstorsjo** [comment]
> @mstorsjo I apologize I thought I had included the link. Here is ld.bfd bug report: https://sourceware.org/bugzilla/show_bug.cgi?id=23872
> The error:
> 
> ```
>   Mingw-w64 runtime failure:
>     Unknown pseudo relocation protocol version 16777216.
> ```

Ah, that one.

I have one potential fix in mind - appending a final object file that contains .rdata contents that is aligned to 4 bytes, included as the last object file in the link command. The object file can be assembled from e.g. this:

```
        .section .rdata$zzzz
        .align 4
        .long 0
```

That should take care to make sure the end of the .rdata section is aligned to 4 bytes, avoiding that particular ld bug.

**Comment by mati865** [comment]
Unfortunately that didn't work. I'll just close it for now.

**Pull request closed without merging**